### PR TITLE
refactor: pipeline abstract cms provider

### PR DESCRIPTION
**Pull request by sfwn**
#### What type of this PR

/kind feature

#### What this PR does / why we need it:

refactor pipeline cms to provider mechanism and use protobuf.

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/erda/workBench/projects/387/issues/all?id=71447&issueFilter__urlQuery=eyJpdGVyYXRpb25JRHMiOls0MzFdLCJzdGF0ZUJlbG9uZ3MiOlsiV09SS0lORyIsIlJFT1BFTiIsIk9QRU4iLCJSRVNPTFZFRCJdLCJhc3NpZ25lZUlEcyI6WyI5MiJdfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6InRhYmxlIiwiY2hpbGRyZW5WYWx1ZSI6eyJrYW5iYW4iOiJwcmlvcml0eSJ9fQ%3D%3D&iterationID=431&type=TASK)

### CODE REVIEW

File: modules/core-services/services/application/appliation.go [hunk 1]
```diff
@@ -769,12 +778,11 @@ func (a *Application) UpdatePublishItemRelations(request *apistructs.UpdateAppPu
...
-			apistructs.PipelineCmsUpdateConfigsRequest{
-				KVs: map[string]apistructs.PipelineCmsConfigValue{"AI": {Value: mk.AI},
-					"AK": {Value: mk.AK}},
-				PipelineSource: apistructs.PipelineSourceDice,
-			}); err != nil {
+		if _, err := a.cms.UpdateCmsNsConfigs(context.Background(), &cmspb.CmsNsConfigsUpdateRequest{
+			Ns:             a.BuildItemMonitorPipelineCmsNs(request.AppID, workspace.String()),
+			PipelineSource: apistructs.PipelineSourceDice.String(),
+			Kvs:            map[string]*cmspb.PipelineCmsConfigValue{"AI": {Value: mk.AI}, "AK": {Value: mk.AK}},
+		}); err != nil {
```
Comment (Effet) [inline modules/core-services/services/application/appliation.go:785]: explain that even struct name and params changed, the reason why

File: modules/dop/services/appcertificate/app_certificate.go [hunk 2]
```diff
@@ -250,10 +259,10 @@ func (c *AppCertificate) ListAllAppCertificates(params *apistructs.AppCertificat
 // PushConfigs 推送Certificate配置到配置管理
 func (c *AppCertificate) PushConfigs(certificatePushReq *apistructs.PushCertificateConfigsRequest) error {
 	var (
-		pushReq = apistructs.PipelineCmsUpdateConfigsRequest{
-			PipelineSource: apistructs.PipelineSourceDice,
+		pushReq = &cmspb.CmsNsConfigsUpdateRequest{
+			PipelineSource: apistructs.PipelineSourceDice.String(),
```
Comment (Effet) [inline modules/dop/services/appcertificate/app_certificate.go:263]: does pb support enum definition?

File: modules/pipeline/providers/cms/provider.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
...41SnUwaEF6Ujg4YXllVmVzTTZRb011Y2cyb0d1Cmh1V1QxNW9LbXlVY2F5dDIrVkNBaVJVZmliNmN3Q3pTSlUyNkFOZk1uWlVqQS83WThQZGcwcFhOSjhuTktiS3EKbUc2dVVlcWdIWENyZjlnTEc4SVRKTVJOaG9VZmJTTjQvNVExMlFtZUFLQlZweitQYTNNR1U5blJUS1luRjVmcApuK3BHK1FLQ0FRRUF3UTMrb2VUMDFFNW5rT0piUStwTEtYMWg3aWloUUsxM0FLdko4dHBCMFRpcVlRTXR0V29JCmJ1QnZJOWZHMTI1UUJxTlVSVTNLN21DT1diNU5YdXdTODZKNjZ6RERkZFA1dkZTUFR3bWJ3TVdkUDJQemtNYXMKUkNsMUJudDJTRGxRV2NLd3Y2S2xrNWZNVm1WWGp3b3VYc2xBWno3MkR5VGU5QmhDMzVQUURVM1R2eVE3aWIwMwo3TWVxVWp3dHZDNmFYTjBaWmlYdWNEWkFMaDlGQnA4cGkyWWZkUzJsellvRGhibVcwV0VITjd2WEFMa3hyYTNHCmZVOW9QeUlMa2JuUG1IQWVIcXlFeTQ4Y3ZGZXZ3Q1RTZXZabElRdEY5U09kRFdaaXZaTFJaZzRxNVd5cHUvaVQKSmUyVnFIeUpJNDZFMkdGZGxXa2JtLzhuckpDdzVwTkZZUUtDQVFCWTNjKzVnY0dkcGJoYkRjenlHNngrcHJVbwpWclVlanNJTUlIZEh1MUUzWW5rWVdrL2hMWEh4anYrbW14RTVOb0hpMmJSTG53VXZhelhReW1CV2l2cW9FRGtmClZsMGR1Tk5xZmpWcG1JY092U1Rkb0wwRHlsZVExT2JEUjBZMEhMZmNvUnJLSUhzOE13b3BWWGtIZjFzamdxVEIKMjNycHd6aDVBTHdaMXJNVDJGNC9CL3RRdlBBMzRTeGlhanQ0RlBHU3pwUE50b0trU1ZOTnM3TS96SXFmY1JtZAppNjNhcW1Tc2pJWEdrZWY1cGpZZTV2WHM2ODZkVldMVGZ4SllMa0N6TkZ4c1dZdXBFaWhZQVJQbnViUGpNd21QCkVWcDBiQTdJNlpnQUQ3Vmw1TkR3UGg4bFZYUlYzWHVyV2dFVXUrcytqMHNTMEZyS2tUeldsRmF1aUhmawotLS0tLUVORCBwcml2YXRlIGtleS0tLS0tCg=="`
```
Comment (Effet) [inline modules/pipeline/providers/cms/provider.go:35]: dont given key in code

File: modules/dop/services/pipeline/pipeline.go [hunk 4]
```diff
@@ -397,18 +407,19 @@ func generatorWorkspaceNS(appID uint64, workspace string) ([]string, error) {
...
-		ns := fmt.Sprintf("%s-%d-%s", apistructs.PipelineAppConfigNameSpacePreFix, appID, branchPrefix)
-		configs, err := p.bdl.GetPipelineCmsNsConfigs(ns, apistructs.PipelineCmsGetConfigsRequest{
+		ns := fmt.Sprintf("%s-%d-%s", cms.PipelineAppConfigNameSpacePrefix, appID, branchPrefix)
+		configs, err := p.cms.GetCmsNsConfigs(context.Background(), &cmspb.CmsNsConfigsGetRequest{
+			Ns:             ns,
 			PipelineSource: "dice",
 		})
 		if err == nil {
-			if len(configs) > 0 {
+			if len(configs.Data) > 0 {
```
Comment (Effet) [inline modules/dop/services/pipeline/pipeline.go:422]: dont know struct of configs, will it panic?

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/core-services/services/application/appliation.go (hunk 1 above)
Comment (sfwn) [reply modules/core-services/services/application/appliation.go:785]: https://developers.google.com/protocol-buffers/docs/style#message_and_field_names

Use underscore_separated_names for field names. So I use `kvs` as field name and generated var name is Kvs, and there is no option to change this generation behaviour.

Now I use `KVs` as field name and set `json_name` option to `kvs`.

File: modules/dop/services/appcertificate/app_certificate.go (hunk 2 above)
Comment (sfwn) [reply modules/dop/services/appcertificate/app_certificate.go:263]: Yes, but it's very hard to use.
And pipelineSource cannot be enumerated for `cdp-` prefixed scenario.

File: modules/dop/services/pipeline/pipeline.go (hunk 4 above)
Comment (sfwn) [reply modules/dop/services/pipeline/pipeline.go:422]: Won't.

File: modules/pipeline/providers/cms/provider.go (hunk 3 above)
Comment (sfwn) [reply modules/pipeline/providers/cms/provider.go:35]: move to pipeline.yaml of provider.

File: modules/dop/services/appcertificate/app_certificate.go (hunk 2 above)
Comment (Effet) [reply modules/dop/services/appcertificate/app_certificate.go:264]: will us support later? then all of this should be comment as TODO

File: conf/pipeline/pipeline.yaml [hunk 5]
```diff
@@ -1 +1,19 @@
...
+
+erda.core.pipeline.cms:
+  base64_encoded_rsa_public_key: "${CMS_BASE64_ENCODED_RSA_PUBLIC_KEY:LS0tLS1CRUdJTiBwdWJsaWMga2V5LS0tLS0KTUlJQ0lqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FnOEFNSUlDQ2dLQ0FnRUFrOCtVK3QyeHhoM1hpREJnRjM2dApxWU5UZmN2NDA4aTdsZnFZRG9TRHMxbDA5bitsLzFOZTQ5b0xxZ0h1ZTQ5MmJHNFI0T0ZHZW1IMktIZmUya3BnCjZpd2tFM0xrZW5KMm56NFdPQWNnOUhiWlA0TFpReGxoeUVwNlE2aHQyekgxZ25Uc2p0QUlzMEZxbXJXZmlVVkQKdFdib1lmSDMvNWZReSs3V00yWkU3bzdnWWxIM1RLR2M5amEvWmgwOTBUZXdULzV3TVhPb1llcFRsWVBmTDVoTwo0em9GeGFpbzltanhpQmVveDNrUkM5RlZsSFM4ZDVlYWRHNkttR2cydjlTaE96SThDaGErRkJHSm83b3E4UEZEClRFMUFuZnBjZml5ckVxVVpzbDZTckl1TjVZUTREM3h1clZnY1RkcG9MV1dpallJbVZ0bytJU3FScW9QemxqVWQKTzdDa2NVRXUvVno2UCt2Vjc4b1JWRktYM0E0aG9vYlFFSkphNlFISmlzN1JQRW5TTjZXS2k4RXkzSlFhT3hXWAppejR3aDk3VmIyZDU4c3l1M0pJSTFOWVlyemtqTitEd1RLV1dqcjVYaVhHSGVCRDFtMmpaMytxV1RCTW1oNC9QCmtWc2M0T29lOG40ZXFoYVc1d2QyaU5jUlRHUS9sUmY4ekNSRlhCN1lvbWJrVlQwc1hVcllXQWFkWURFUEFmazUKTncvUjJaTXkyNGVhd0ZCcTVmYVB6VVJWRUY4WC9uUm5kL1YwUFZBSGgySG9CeFJaZzFkSGJrSWQ3SUo5R2cxbwpKVzJZOTlobzRpK0QvTDl2cWNPOVRyOXN0dStWcG1UQ1BRdFZqWHlpY0FuZmN4MWxhOEI0Q2Y4azhWN1RBSmJWCm14SjdaUTJEbGs3TTdBYzNTamVEUmJrQ0F3RUFBUT09Ci0tLS0tRU5EIHB1YmxpYyBrZXktLS0tLQo=}"
```
Comment (Effet) [inline conf/pipeline/pipeline.yaml:18]: should not embed key in config.yml, if changes hard you can comment when we refactor it.

File: modules/core-services/services/application/appliation.go [hunk 6]
```diff
@@ -769,11 +779,11 @@ func (a *Application) UpdatePublishItemRelations(request *apistructs.UpdateAppPu
 func (a *Application) PipelineCmsConfigRequest(request *apistructs.UpdateAppPublishItemRelationRequest) error {
 	for workspace, mk := range request.AKAIMap {
 		// bundle req
-		if err := a.bdl.CreateOrUpdatePipelineCmsNsConfigs(a.BuildItemMonitorPipelineCmsNs(request.AppID, workspace.String()),
-			apistructs.PipelineCmsUpdateConfigsRequest{
-				KVs: map[string]apistructs.PipelineCmsConfigValue{"AI": {Value: mk.AI},
-					"AK": {Value: mk.AK}},
-				PipelineSource: apistructs.PipelineSourceDice,
+		if _, err := a.cms.UpdateCmsNsConfigs(apis.WithInternalClientContext(context.Background(), "core-services"),
```
Comment (Effet) [inline modules/core-services/services/application/appliation.go:782]: what different between CreateOrUpdatePipelineCmsNsConfigs and UpdateCmsNsConfigs?

File: modules/dop/services/appcertificate/app_certificate.go (hunk 2 above)
Comment (sfwn) [reply modules/dop/services/appcertificate/app_certificate.go:264]: No.

Comment (sfwn) [reply modules/dop/services/appcertificate/app_certificate.go:264]: `PipelineSource` is not an Enum, **CANNOT** be enumerated.

File: modules/dop/endpoints/configmanage.go [hunk 7]
```diff
@@ -347,12 +352,14 @@ func (e *Endpoints) generatorPipelineNS(appID uint64) (*apistructs.PipelineConfi
...
 		if branchPrefix == gitflowutil.HOTFIX_WITHOUT_SLASH || branchPrefix == gitflowutil.SUPPORT_WITHOUT_SLASH {
-			ns := fmt.Sprintf("%s-%d-%s", apistructs.PipelineAppConfigNameSpacePreFix, appID, branchPrefix)
-			configs, err := e.bdl.GetPipelineCmsNsConfigs(ns, apistructs.PipelineCmsGetConfigsRequest{
-				PipelineSource: "dice",
-			})
+			ns := fmt.Sprintf("%s-%d-%s", cms.PipelineAppConfigNameSpacePrefix, appID, branchPrefix)
+			configs, err := e.pipelineCms.GetCmsNsConfigs(utils.WithInternalClientContext(context.Background()),
+				&cmspb.CmsNsConfigsGetRequest{
+					Ns:             ns,
+					PipelineSource: "dice",
```
Comment (Effet) [inline modules/dop/endpoints/configmanage.go:359]: this hard-coded "dice" should be const or refactor it out.

if cannot refactor at this time, comment TODO

File: modules/dop/services/autotest/global_config.go [hunk 8]
```diff
@@ -231,22 +237,22 @@ func (svc *Service) createOrUpdatePipelineCmsGlobalConfigs(cfg *apistructs.AutoT
...
+			Type:        cms.ConfigTypeKV,
+			Operations:  &cms.DefaultOperationsForKV,
 			Comment:     "auto test api global config",
 			From:        apistructs.PipelineSourceAutoTest.String(),
 		}
 
 		// 转成 config.autotest.xx 语法由 pipeline 渲染
 		for _, item := range cfg.APIConfig.Global {
-			kvs[apistructs.PipelineSourceAutoTest.String()+"."+item.Name] = apistructs.PipelineCmsConfigValue{
+			kvs[apistructs.PipelineSourceAutoTest.String()+"."+item.Name] = &cmspb.PipelineCmsConfigValue{
```
Comment (Effet) [inline modules/dop/services/autotest/global_config.go:251]: is this key pattern any description? you should comment it

File: modules/dop/utils/internal_ctx.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+package utils
+
+import (
+	"context"
+
+	"github.com/erda-project/erda/pkg/common/apis"
+)
+
+func WithInternalClientContext(ctx context.Context) context.Context {
+	return apis.WithInternalClientContext(ctx, "dop")
```
Comment (Effet) [inline modules/dop/utils/internal_ctx.go:23]: bad hard-coded "dop", you should get it in a-native-way, quite like get module name by erda-infra's ability.

File: modules/pipeline/providers/cms/pipeline_cms.go [hunk 10]
```diff
@@ -251,26 +254,39 @@ func (c *pipelineCm) GetConfigs(ctx context.Context, ns string, globalDecrypt bo
...
+
 func getPipelineSourceFromContext(ctx context.Context) (apistructs.PipelineSource, error) {
-	pipelineSource, ok := ctx.Value(CtxKeyPipelineSource).(apistructs.PipelineSource)
-	if !ok || pipelineSource == "" {
+	var source apistructs.PipelineSource
+	switch ctx.Value(CtxKeyPipelineSource).(type) {
+	case apistructs.PipelineSource:
+		source = ctx.Value(CtxKeyPipelineSource).(apistructs.PipelineSource)
+	case string:
+		source = apistructs.PipelineSource(ctx.Value(CtxKeyPipelineSource).(string))
```
Comment (Effet) [inline modules/pipeline/providers/cms/pipeline_cms.go:284]: what if neither apistructs.PipelineSource nor string? if imposible, you should comment reason in code

**Review by Effet — CHANGES_REQUESTED** [review]

File: conf/pipeline/pipeline.yaml (hunk 5 above)
Comment (sfwn) [reply conf/pipeline/pipeline.yaml:18]: Later will use kms to make key-change operation easier.

File: modules/core-services/services/application/appliation.go (hunk 6 above)
Comment (sfwn) [reply modules/core-services/services/application/appliation.go:782]: It's the same. `UpdateCmsNsConfigs` is the original endpoint api name.

File: modules/dop/endpoints/configmanage.go (hunk 7 above)
Comment (sfwn) [reply modules/dop/endpoints/configmanage.go:359]: Adjusted it by the way.

File: modules/dop/services/autotest/global_config.go (hunk 8 above)
Comment (sfwn) [reply modules/dop/services/autotest/global_config.go:251]: It's dop side code. I won't care about inner logic in the PR.

File: modules/dop/utils/internal_ctx.go (hunk 9 above)
Comment (sfwn) [reply modules/dop/utils/internal_ctx.go:23]: You are right. I add the TODO comment and after all dop services migrated to providers and passed with ctx.

File: modules/pipeline/providers/cms/pipeline_cms.go (hunk 10 above)
Comment (sfwn) [reply modules/pipeline/providers/cms/pipeline_cms.go:284]: add default branch.

**Review by Effet — APPROVED** [review]

**Pull request merged**